app/lib/admin/backend.dart

```diff
@@ -0,0 +1,200 @@
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:client_data/admin_api.dart' as api;
+import 'package:continuation_token/continuation_token.dart';
+import 'package:gcloud/db.dart';
+import 'package:gcloud/service_scope.dart' as ss;
+import 'package:logging/logging.dart';
+
+import '../account/backend.dart';
+import '../account/models.dart';
+import '../package/models.dart';
+import '../publisher/models.dart';
+import '../shared/configuration.dart';
+import '../shared/exceptions.dart';
+
+final _logger = Logger('pub.admin.backend');
+final _continuationCodec = stringContinuationCodec();
+
+/// Sets the admin backend service.
+void registerAdminBackend(AdminBackend backend) =>
+    ss.register(#_adminBackend, backend);
+
+/// The active admin backend service.
+AdminBackend get adminBackend => ss.lookup(#_adminBackend) as AdminBackend;
+
+/// Represents the backend for the admin handling and authentication.
+class AdminBackend {
+  final DatastoreDB _db;
+  AdminBackend(this._db);
+
+  Future<R> _withAdmin<R>(Future<R> fn(AuthenticatedUser user)) async {
+    return await withAuthenticatedUser((u) async {
+      // TODO: remove once [withAuthenticatedUser] is using [User]
+      final user = await accountBackend.lookupUserById(u.userId);
+      if (!activeConfiguration.admins.contains(user.oauthUserId)) {
+        throw AuthorizationException.userIsNotAdminForPubSite();
+      }
+      return await fn(u);
+    });
+  }
+
+  /// List users.
+  Future<api.AdminListUsersResponse> listUsers(
+      {String continuationToken, int limit = 1000}) async {
+    InvalidInputException.checkRange(limit, 'limit', minimum: 1, maximum: 1000);
+    return await _withAdmin((user) async {
+      final query = _db.query<User>()
+        ..order('__key__')
+        ..limit(limit);
+
+      if (continuationToken != null) {
+        String lastId;
+        try {
+          lastId = _continuationCodec.decode(continuationToken);
+        } on FormatException catch (_) {
+          throw InvalidInputException.continuationParseError();
+        }
+        InvalidInputException.checkNotNull(lastId, 'continuationToken');
+
+        // NOTE: we should fix https://github.com/dart-lang/gcloud/issues/23
+        //       and remove the toDatastoreKey conversion here.
+        final key =
+            _db.modelDB.toDatastoreKey(_db.emptyKey.append(User, id: lastId));
+        if (lastId != null) {
+          query.filter('__key__ >', key);
+        }
+      }
+
+      final users = await query.run().toList();
+      // We may return a page with users less then a limit, but we always
+      // set the continuation token to the correct value.
+      final newContinuationToken = users.length < limit
+          ? null
+          : _continuationCodec.encode(users.last.userId);
+      users.removeWhere((u) => u.isDeleted);
```

Comment:
Can't we filter on this in the query? if we add `isDeleted` to the index?

Reply:
I'm not sure if we can filter on values that are `null`. I thought those entities are not indexed at all.

Reply:
Should we just make a utility that adds `isDeleted` to all User entities and then have the integrity_check verify this as well?

Essentially, I'm suggesting that we make it an invariant that `isDeleted` is set to either true or false.

Reply:
Yeah, let's do this in a separate PR.